solr/core/src/java/org/apache/solr/handler/QueryRegisterHandler.java

```diff
@@ -0,0 +1,153 @@
+import org.apache.solr.handler.component.SearchComponent;
+import org.apache.solr.request.SolrQueryRequest;
+import org.apache.solr.response.SolrQueryResponse;
+import org.apache.solr.update.processor.MonitorUpdateProcessorFactory;
+import org.apache.solr.util.plugin.SolrCoreAware;
+import org.apache.zookeeper.KeeperException;
+import org.jose4j.json.JsonUtil;
+import org.jose4j.lang.JoseException;
+
+public class QueryRegisterHandler extends RequestHandlerBase implements SolrCoreAware {
+
+  public static final String PARSER_DEFAULT_FIELD_NAME = "defaultField";
+  public static final String PARAM_QUERY_ID_NAME = "id";
+  public static final String PARAM_COLLECTION_NAMES = "collections";
+  public static final String ZK_KEY_SOLR_PARAMS = "params";
+  public static final String ZK_KEY_QUERY_STRING = "q";
+  public static final String ZK_KEY_VERSION = "version";
+
+  private SolrCore core;
+
+  static Query parse(String query) {
+    try {
+      StandardQueryParser parser = new StandardQueryParser();
+      return parser.parse(query, PARSER_DEFAULT_FIELD_NAME);
+    } catch (QueryNodeException e) {
+      throw new IllegalArgumentException(e);
+    }
+  }
+
+  @Override
+  public void handleRequestBody(SolrQueryRequest req, SolrQueryResponse rsp) throws Exception {
+    CoreContainer cc = req.getCore().getCoreContainer();
+    SolrZkClient client = cc.getZkController().getZkClient();
+    SolrParams params = req.getParams();
+    String queryId = params.get(PARAM_QUERY_ID_NAME);
+
+    // use query component to parse query string
+    List<SearchComponent> components = new ArrayList<>();
+    components.add(core.getSearchComponent("query"));
+    ResponseBuilder rb = new ResponseBuilder(req, rsp, components);
+    for (SearchComponent c : components) c.prepare(rb);
+    Query query = rb.getQuery();
+
+    // serialize SolrParams
+    byte[] paramBytes = getBytes(params);
+    String paramString = Base64.byteArrayToBase64(paramBytes);
+
+    registerQueryToZk(client, queryId, query, paramString);
+  }
+
+  /**
+   * Write the query to Zookeeper.
+   * All queries are stored in a single node, so read -> modify -> write.
+   * "a single handler instance is reused for all relevant queries" -> synchronized method
+   *
+   * @param client      Zookeeper client
+   * @param queryId     The user specified id of query
+   * @param query       Parsed Lucene {@link Query} object
+   * @param paramString The string(base64) representation of the serialized query
+   * @throws KeeperException      If an exception occurs related to Zookeeper
+   * @throws InterruptedException If thread is interrupted
+   * @throws JoseException        If an exception occurs in JsonUtil
+   */
+  synchronized private void registerQueryToZk(SolrZkClient client, String queryId, Query query, String paramString) throws KeeperException, InterruptedException, JoseException {
+    String path = MonitorUpdateProcessorFactory.zkQueryPath;
+    if (!client.exists(path, true)) {
```

Comment:
Moving these code to handler.init() caused some exceptions and make it fail to create cores. So I added a flag variable here to avoid unnecessary zk calls.